larray/tests/test_session.py

```diff
@@ -103,173 +113,142 @@ def test_setitem(session):
+    assertObjListEqual(session.filter(kind=Array), [e, g, f, h])
     assert list(session.filter(kind=dict)) == [{}]
-    assert list(session.filter(kind=(Axis, Group))) == [b, b12, a, a01]
+    assert list(session.filter(kind=(Axis, Group))) == [b, b024, a, a2, anonymous, a01, ano01]
 
 
 def test_names(session):
-    assert session.names == ['a', 'a01', 'b', 'b12', 'c', 'd', 'e', 'f', 'g']
+    assert session.names == ['a', 'a01', 'a2', 'ano01', 'anonymous', 'b', 'b024',
+                             'c', 'd', 'e', 'f', 'g', 'h']
     # add them in the "wrong" order
     session.add(i='i')
-    session.add(h='h')
-    assert session.names == ['a', 'a01', 'b', 'b12', 'c', 'd', 'e', 'f', 'g', 'h', 'i']
+    session.add(j='j')
+    assert session.names == ['a', 'a01', 'a2', 'ano01', 'anonymous', 'b', 'b024',
+                             'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
 
 
-def test_h5_io(tmpdir, session, meta):
-    fpath = tmp_path(tmpdir, 'test_session.h5')
+def _test_io(fpath, session, meta, engine):
+    is_excel_or_csv = 'excel' in engine or 'csv' in engine
+
+    kind = Array if is_excel_or_csv else (Axis, Group, Array)
+    session = session.filter(kind=kind)
+
     session.meta = meta
-    session.save(fpath)
 
+    # save and load
+    session.save(fpath, engine=engine)
     s = Session()
-    s.load(fpath)
-    # HDF does *not* keep ordering (ie, keys are always sorted +
-    # read Axis objects, then Groups objects and finally Array objects)
-    assert list(s.keys()) == ['a', 'b', 'a01', 'b12', 'e', 'f', 'g']
-    assert s.meta == meta
+    s.load(fpath, engine=engine)
+    # use Session.names instead of Session.keys because CSV, Excel and HDF do *not* keep ordering
+    assert s.names == session.names
+    assert s.equals(session)
+    if not PY2 and not is_excel_or_csv:
+        for key in s.filter(kind=Axis).keys():
+            assert s[key].dtype == session[key].dtype
+    if engine != 'pandas_excel':
+        assert s.meta == meta
 
     # update a Group + an Axis + an array (overwrite=False)
-    a2 = Axis('a=0..2')
-    a2_01 = a2['0,1'] >> 'a01'
-    e2 = ndtest((a2, 'b=b0..b2'))
-    Session(a=a2, a01=a2_01, e=e2).save(fpath, overwrite=False)
+    a3 = Axis('a=0..3')
+    a3_01 = a3['0,1'] >> 'a01'
+    e2 = ndtest((a3, 'b=b0..b2'))
+    Session(a=a3, a01=a3_01, e=e2).save(fpath, overwrite=False, engine=engine)
     s = Session()
-    s.load(fpath)
-    assert list(s.keys()) == ['a', 'b', 'a01', 'b12', 'e', 'f', 'g']
-    assert s['a'].equals(a2)
-    assert all(s['a01'] == a2_01)
+    s.load(fpath, engine=engine)
+    if engine == 'pandas_excel':
+        # Session.save() via engine='pandas_excel' always overwrite the output Excel files
+        assert s.names == ['e']
+    elif is_excel_or_csv:
+        assert s.names == ['e', 'f', 'g', 'h']
+    else:
+        assert s.names == session.names
+        assert s['a'].equals(a3)
+        if engine == 'pickle':
```

Comment:
Why does pickle needs a different code? I don't see any reason it **should**. Isn't this masking a bug? If this is not a bug, there should at least be a comment explaining the situation/why this difference in treatment is necessary.

Reply:
The reason is that the key of a group is converted as a Numpy ndarray when using the HDF format while the pickle format keeps the original type of the key (in this case: a simple list). 
But the unittest function is fixed by by using the Group.equals() method.